[SimplifyCFG] Forward indirect switch condition value if it can help fold the PHI

dianqk opened the pull request:
Fixes #95919.

dianqk committed:
Avoid affecting unrelated test cases

dianqk committed:
Pre-commit test cases

dianqk committed:
[SimplifyCFG] Don't limit the number of simultaneous forwards from switch condition

nikic reviewed (commented):
Is it possible to also add a test case where this shows an actual benefit like in https://github.com/llvm/llvm-project/issues/95919? Or does this require multiple passes?

DianQK commented:
> Is it possible to also add a test case where this shows an actual benefit like in #95919? Or does this require multiple passes?

Yes, the existing `-switch-range-to-icmp` is already sufficient.

dtcxzyw commented on llvm/test/CodeGen/AArch64/arm64-jumptable.ll:
```diff
@@ -18,7 +18,7 @@ bb3:
   br label %exit.sink.split
 exit.sink.split:
-  %.sink = phi i32 [ 5, %bb4 ], [ %b, %bb1 ], [ 3, %bb3 ], [ %a, %entry ]
+  %.sink = phi i32 [ 5, %bb4 ], [ %b, %bb1 ], [ 7, %bb3 ], [ %a, %entry ]
```
Can you explain this change?

DianQK replied:
Because `%bb3` happens to meet this optimization condition, changing it to another value can prevent this test case from being affected.
See:
https://github.com/llvm/llvm-project/blob/bdc0afc87181d4f7ab8aad2da6fa70a1204f0a84/llvm/test/CodeGen/AArch64/arm64-jumptable.ll#L10-L21

dtcxzyw reviewed (approved):
LGTM. Thank you!

nikic reviewed (approved):
LGTM

zmodem commented:
> To me, this limit doesn't make sense.

Looks like I added this in 4ab4a8e63a5a99face7d7b523529d3805385c93a

I think the point of the `Indexes.size()` check is explained by the "if that would mean that some of the destination blocks of the switch can be folded away" part of this comment:

https://github.com/llvm/llvm-project/blob/9cbedd925cba9e8ef76c50caa6d6ab4b0cc79c8f/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L5774-L5776

The point of the forwarding is to make *multiple blocks* produce the same value for the phi node (the switch condition), so that they can be folded into one block.

If we're forwarding the condition value for just a single block, that's just replacing a constant with a non-constant which seems counter-productive in general.

Looking at the example in #95919

> ```
> bb3:                                              ; preds = %bb7, %bb2, %bb
>  ...
>  %i4 = phi i64 [ 1, %bb7 ], [ 1, %bb2 ], [ %arg1, %bb ]
> ```
> We can replace `[ 1, %bb2 ] with [ %arg1, %bb2 ]`.

What makes that beneficial is that the incoming value for `%bb` is already `%arg1`. That makes me think the right fix is rather to take that into account in `FindPHIForConditionForwarding`, something like:

```
diff --git a/llvm/lib/Transforms/Utils/SimplifyCFG.cpp b/llvm/lib/Transforms/Utils/SimplifyCFG.cpp
index 4e2dc7f2b2f4..980dee7707ac 100644
--- a/llvm/lib/Transforms/Utils/SimplifyCFG.cpp
+++ b/llvm/lib/Transforms/Utils/SimplifyCFG.cpp
@@ -5743,7 +5743,7 @@ static bool eliminateDeadSwitchCases(SwitchInst *SI, DomTreeUpdater *DTU,
 /// by an unconditional branch), look at the phi node for BB in the successor
 /// block and see if the incoming value is equal to CaseValue. If so, return
 /// the phi node, and set PhiIndex to BB's index in the phi node.
-static PHINode *FindPHIForConditionForwarding(ConstantInt *CaseValue,
+static PHINode *FindPHIForConditionForwarding(Value *SwitchCondition, ConstantInt *CaseValue,
                                               BasicBlock *BB, int *PhiIndex) {
   if (BB->getFirstNonPHIOrDbg() != BB->getTerminator())
     return nullptr; // BB must be empty to be a candidate for simplification.
@@ -5761,7 +5761,7 @@ static PHINode *FindPHIForConditionForwarding(ConstantInt *CaseValue,
     assert(Idx >= 0 && "PHI has no entry for predecessor?");
 
     Value *InValue = PHI.getIncomingValue(Idx);
-    if (InValue != CaseValue)
+    if (InValue != CaseValue && InValue != SwitchCondition)
       continue;
 
     *PhiIndex = Idx;
@@ -5811,7 +5811,7 @@ static bool ForwardSwitchConditionToPHI(SwitchInst *SI) {
 
     // Collect phi nodes that are indirectly using this switch's case constants.
     int PhiIdx;
-    if (auto *Phi = FindPHIForConditionForwarding(CaseValue, CaseDest, &PhiIdx))
+    if (auto *Phi = FindPHIForConditionForwarding(SI->getCondition(), CaseValue, CaseDest, &PhiIdx))
       ForwardingNodes[Phi].push_back(PhiIdx);
   }
```

DianQK commented:
Thank you for your explanation. However, I still think is meaningful.

I've chosen some examples to demonstrate the final results in assembly code. Analyzing more practical changes might be challenging, so I'm comparing the number of instructions, which might not be ideal but should at least show that this transformation is not harmful.

- I replaced `undef` with other values, and this transformation resulted in a slight increase in instructions: https://alive2.llvm.org/ce/z/2yE6g_, but, on arm64, the instruction count remains the same.: https://alive2.llvm.org/ce/z/zMVH2_
- I continued by modifying the switch conditions with other values, which reduced some instructions: https://alive2.llvm.org/ce/z/y2Xa8k (Similarly for arm64: https://alive2.llvm.org/ce/z/mfKixU)
- I removed some code, which I believe is an improvement: https://alive2.llvm.org/ce/z/DfLwMV
- Another similar example also shows improvement: https://alive2.llvm.org/ce/z/hwJheB

I suspect the main optimization here comes from register reuse. I'm not sure how this would perform within a large function body, as it might complicate instruction selection.

For me, another important point is that *we can easily revert this forward if needed for other Passes. However, turning constants back into variables requires the dominance information.*

zmodem commented:
(I can't access the alive2 links, the server seems to be down?)

I don't have any strong opinion about whether always "forwarding" the switch condition is right or wrong. I suppose it comes down to what the following passes are best able to handle, and @nikic is probably the right person to answer that.

On the one hand, it seems bad to replace constants with a variable -- that's why 4ab4a8e was conservative and only does it when it's likely to cause blocks to get folded. On the other hand, as you say, it's easy to propagate such constants later.

I do think my diff above would fix your use case? So if we do decide to "always forward" I think we need some motivation, and we should update the comment for `FindPHIForConditionForwarding` with that motivation rather than talk about folding away destination blocks.

nikic commented:
It's a tricky question. I'd generally say that having constants instead of variables is "more canonical" -- and in fact, we have quite a lot of passes that will undo what SimplifyCFG does here. At least IPSCCP, CVP and GVN do this. This is bad in itself, because it means that IR toggles back and forth between two forms. Conversely, it also means that many passes have a chance to optimize IR in both forms. The backend generally prefers phi operands not to be constants (though this is a pretty generic problem and SimplifyCFG isn't really the place to address it I think).

DianQK commented:
> (I can't access the alive2 links, the server seems to be down?)

Fixed. GitHub seems to not recognize `_`.

> I do think my diff above would fix your use case? So if we do decide to "always forward" I think we need some motivation, and we should update the comment for `FindPHIForConditionForwarding` with that motivation rather than talk about folding away destination blocks.

I think so, but it would also make this part of the code a bit more complex. In fact, I don't have a clear choice for the change here.

DianQK commented:
> It's a tricky question. I'd generally say that having constants instead of variables is "more canonical" -- and in fact, we have quite a lot of passes that will undo what SimplifyCFG does here. At least IPSCCP, CVP and GVN do this. This is bad in itself, because it means that IR toggles back and forth between two forms. Conversely, it also means that many passes have a chance to optimize IR in both forms. The backend generally prefers phi operands not to be constants (though this is a pretty generic problem and SimplifyCFG isn't really the place to address it I think).

I have seen some similar things. These transforms are just back and forth, and it doesn't seem like they are being utilized.
I think a possible solution could be to treat this as a result of some analysis pass, and apply it when it's beneficial to use this result for certain transformations. Then transform it to phi operands before the backend.

DianQK commented:
In this PR, I will change it to forward when it finds phi folding, which aligns more with SimplifyCFG behavior.

dianqk committed:
Check if it helps to fold PHI

DianQK commented:
This patch will cause values in the PHI to become switch condition values, leading to an infinite loop. Additionally, it doesn't align well with the intent of the function. I have changed it to check if the PHI already contains the switch condition value.

Forwarding is meaningful for the test case in this PR: [https://alive2.llvm.org/ce/z/Z46qt6](https://alive2.llvm.org/ce/z/Z46qt6).
Based on the above discussions, I plan to implement the following changes in subsequent PRs:

- Forward switch condition value to constant before backend execution
- Forward only when the PHI can be folded. For instance, when there are multiple PHI nodes, checking only itself is insufficient. This should also help avoid potentially pointless transforms with IPSCCP and other passes, possibly reducing compilation time slightly.

DianQK commented:
@zmodem Any more comments? :)

zmodem reviewed (approved):
lgtm

vitalybuka commented:
https://lab.llvm.org/buildbot/#/builders/72/builds/483 is likely flake

DianQK commented:
> https://lab.llvm.org/buildbot/#/builders/72/builds/483 is likely flake

😄 I should wait for the next result.
